Additional validation on instance group configs in the role manifest

mook-as opened the pull request:
This PR does a few things:

- Converts the `app/validation` tests to be driven by the YAML files, to make it easier to add more tests
- Add validation of instance-group-specific configuration templates (including variable use and sorting)
- Change the validation of global configuration templates to only allow things that are in actually used jobs (as opposed to any job in any referenced release)
- Move some tests from `model/resolver` to `app/validation`; this makes it possible to load not-quite-valid manifests. We should eventually move validation (but not normalization) out of the resolver.
- Make `ErrorList` actually implement `error`, so that we can print it out easier (without having convert).

Note that this will require changes to the UAA & SCF role manifests to conform to the new checks.

mook-as committed:
validation/errors.go: Make ErrorList implement error interface

Also, add a method to return the list of errors as a string slice.  This
lets us check that a specific expected error is in the list (instead of
just relying on substring matches and hoping it works).

mook-as committed:
app/validation: convert TestValidation to be data-driven

This should let us split up the test easier, and to add more self-
contained tests.

mook-as committed:
app/validation_test: split large validation test

Split the one large file with the various issues into a bunch of
separate ones to make it easier to see what issue is

mook-as committed:
app/validation: only allow properties in used jobs

Make it so that when we check for properties, we only allow the ones in
jobs that are used (and not just in any release, regardless of whether
the job the property is defined in is used).

mook-as committed:
app/validation_test: move 2 more tests to be data-driven

This moves TsetValidationOk and TestValidationHash.

mook-as committed:
app/validation: check that instance-group templates are sorted

mook-as committed:
model/resolver: move checking for global template sorting to app

This moves the check for templates being sorted at the global level from
model/resolver (when the manifest is loaded) to app/validation (after
the manifest is loaded).  This way we can share work at the later stage,
and we can have better debugging if things go wrong.

mook-as committed:
model: make .Configuration.Template a parsed structure

This changes things so that we parse .Configuration.Template (so it's a
structure instead of a raw yaml.MapSlice), and move some validation from
model/resolver/ into app/.

mook-as committed:
model/resolver: move variable sorting validation to app/

This pushes non-essential validation later.

mook-as committed:
app/validation: add test case for extending instance group hashes

mook-as committed:
app/validation: include builtin variables

Ensure that builtins are still counted as defined.

mook-as committed:
model/configuration: Customize ConfigurationTemplate serialization

Ensure that we create the same output in env2conf files on the docker
image for consumption by configgin.

andreas-kupries commented on model/resolver/resolver.go:
```diff
@@ -609,3 +602,21 @@ clusterRoleLoop:
 	return errors
 }
+
+// calculateConfigurationTemplates caculates the global configuration templates
```
nit: `caculates` vs `calculates`

mook-as replied:
Fixed in 311c0c12dbf55e61bd2f9dc2332075f7a8afc131

andreas-kupries reviewed (changes requested):
Typo.

f0rmiga commented on app/validation.go:
```diff
@@ -48,28 +63,110 @@ func (f *Fissile) Validate() validation.ErrorList {
+
+		// Collect all property defaults across all instance groups for a global
+		// check later.
+		for propertyName, defaults := range propertyDefaults {
```
Could we pull this out into a function that we can share with https://github.com/cloudfoundry-incubator/fissile/pull/496/files#diff-d869de81fbd78443594a015394c848c7R314?

mook-as replied:
Fixed in 129e316b441ccd8d7603edb14ed8d4aeb0040238: We don't actually need this at all, we're just doing the same work twice. Thanks for spotting that.

andreas-kupries commented on app/validation_test.go:
```diff
@@ -2,64 +2,96 @@ package app
-		// `XXX`, // Trigger a fail which shows the contents of `actual`. Also template for new assertions.
+	require.NoError(t, err)
+
+	roleManifestDir, err := os.Open(filepath.Join(workDir, "../test-assets/role-manifests/app/validation/"))
```
I had to go looking at the files in that directories to start understanding how you have made it data driven, with their sections essentially describing pieces or role manifest, opinions, and the errors expected from the validation code.

Maybe add a basic comment here to give an overview of things ?

mook-as replied:
Comment added (at the top of the function) in 311c0c12dbf55e61bd2f9dc2332075f7a8afc131; let me know if that's what you're looking for.

f0rmiga commented on app/validation.go:
```diff
@@ -188,30 +308,22 @@ func checkForDuplicatesBetweenManifestAndLight(light map[string]string, roleMani
+					// different role
+					continue
+				}
+				check[property] = struct{}{}
```
Every time I see `struct{}{}` I find it to be the ugliest thing, but at the same time, I love what it does.

andreas-kupries reviewed (commented):
Code may require a comment

mook-as committed:
Address review comments:

https://github.com/cloudfoundry-incubator/fissile/pull/496#discussion_r279949236
https://github.com/cloudfoundry-incubator/fissile/pull/496#discussion_r279952581

mook-as committed:
app/validation: remove duplicated code

Addresses
https://github.com/cloudfoundry-incubator/fissile/pull/496#discussion_r279950142

f0rmiga commented on app/validation.go:
```diff
@@ -122,39 +203,62 @@ func checkParentsOfUndefined(p string, bosh propertyDefaults) bool {
+
+// checkForUndefinedVariables checks that all configuration templates are
+// defined in the variables section
+func checkForUndefinedVariables(label string, templates map[string]model.ConfigurationTemplate, variableUsage map[string]int, isGlobal bool) validation.ErrorList {
```
For this one, we may want it as a method of a `validator` struct that initializes and hols the state of `variableUsage`. Also, move the `range templates` to the caller, as this function don't need the whole thing (law of Demeter, or principle of least knowledge). Then we eliminate the need for passing `isGlobal` as well.

mook-as replied:
Changed in 9b6136ede07c51872f28e995c24e15c78b4978df

f0rmiga commented on app/validation.go:
```diff
@@ -48,28 +63,94 @@ func (f *Fissile) Validate() validation.ErrorList {
+}
+
+// checkForUndefinedProperties checks that all properties are known
+func checkForUndefinedProperties(label string, global bool, templates map[string]model.ConfigurationTemplate, knownProperties model.PropertyDefaults) validation.ErrorList {
```
Here, we can make this function a method of a `validator` struct that takes the `knownProperties` as part of the initialization, and the `range templates` can happen on the caller side, so we don't need to pass the whole templates, neither the `global` parameter.

mook-as replied:
Changed in 9b6136ede07c51872f28e995c24e15c78b4978df

mook-as committed:
app/validation: address review comments

Addresses:
https://github.com/cloudfoundry-incubator/fissile/pull/496#discussion_r279970120
https://github.com/cloudfoundry-incubator/fissile/pull/496#discussion_r279970840

Move the validation state to a struct, to avoid passing opinions etc.
repeatedly.

Also:
- Use a channel for error reporting, to avoid repeated append() calls
- Add test for constant manifest entry duplicating opinions
- Add check for invalid mustache template expansion
- Inline checkForDuplicateProperty()

f0rmiga commented on app/validation.go:
```diff
@@ -6,109 +6,198 @@ import (
+	return &validator{
+		errOut:           errOut,
+		f:                f,
+		propertyDefaults: f.collectPropertyDefaults(),
```
I don't think this belongs to the validator. It's weird that we have it being passed to `checkForUndefinedProperty`, e.g.

We can call `f.collectPropertyDefaults()` and then pass the result to `checkForUndefinedProperty`.

And for `checkLightDefaults`, we pass it as a parameter.

mook-as replied:
Fixed in a7c161c2d8fa267f03009f7b9f063996b53c8e4b (didn't realize we only use it outside `validate()` in one place).

f0rmiga commented on app/validation.go:
```diff
@@ -6,109 +6,198 @@ import (
+			v.checkForUndefinedProperty(
+				"configuration.templates",
+				propertyName,
+				v.propertyDefaults)
```
This is the weird part.

mook-as replied:
This is because we call `v.checkForUndefinedProperty` elsewhere, with the instance-group-specific properties instead. I believe this is the only validation function we do this.

mook-as replied:
Fixed in a7c161c2d8fa267f03009f7b9f063996b53c8e4b

f0rmiga commented on app/validation.go:
```diff
@@ -6,109 +6,198 @@ import (
 
 	// No dark opinions must have defaults in light opinions
-	allErrs = append(allErrs, checkForDarkInTheLight(darkOpinions, lightOpinions)...)
+	v.checkForDarkInTheLight()
```
> And the light shineth in darkness; and the darkness comprehended it not.

mook-as committed:
app/validation: move propertyDefaults out of struct validator

It's only used in two functions; just make validate() pass it in to
v.checkLightDefaults() instead.
